Python3 convert braille drivers

feerrenrut opened the pull request:
<!--
Please fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/wiki/Github-pull-request-template-explanation-and-examples
-->

### Link to issue number:
Related to #9637

### Summary of the issue:
The braille driver framework, and many of the drivers use strings interchangeably with bytes.

### Description of how this pull request fixes the issue:
Since quite a careful approach is required to track down which literals should be bytes / vs strings and much of the surrounding logic needs to be modified I am converting the braille drivers to python 3 files-wise rather than feature-wise as we are elsewhere in the code base.

To help with this I have been using typehints and a linter to find issues. This is not perfect, it's quite likely issues have slipped through.  

### Testing performed:
None, this will require testing from people with devices.

Several people have voluntered to test with different devices:
- alva
- baum
  -  @leonardder 
- BrailleNote
  - @josephsl 
- BrailliantB
- braltty
- ecoBraille
- euroBraille
   - @leonardder
  - Esys 24
    - @zstanecic
- FreedomScientific
- handy tech
  - @leonardder 
- hims
  -  @leonardder 
- lili
- papenmeier
- papenmeier serial
- seika
- superBrl

### Known issues with pull request:

### Change log entry:

Section: New features, Changes, Bug fixes

feerrenrut committed:
fix malformed newline character

feerrenrut committed:
Python3: Convert braille driver framework.

Focusing specifically on bytes/str

feerrenrut committed:
Python3: Convert alva driver

feerrenrut committed:
Python3: Convert baum driver

feerrenrut committed:
Python3: Convert BrailleNote driver

feerrenrut committed:
Python3: Convert BrailliantB driver

feerrenrut committed:
Python3: Convert ecoBraille driver

feerrenrut committed:
Python3: Convert eurobraille driver

feerrenrut committed:
Python3: Convert seika driver

feerrenrut committed:
Python3: Convert superBrl driver

feerrenrut committed:
Python3: Convert freedomScientific driver

feerrenrut committed:
Python3: Convert handyTech driver

leonardder commented:
I can test the hims, eurobraille, handy tech and baum drivers. Furthermore, I wrote most of the ALVA driver so will check that from a code point of view.

josephsl commented:
Hi, I can ask my BrailleNote to take part in testing this (since I also wrote code for Apex and QT support). Thanks.

 

From: Leonard de Ruijter <notifications@github.com> 
Sent: Thursday, June 13, 2019 7:52 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Subject: Re: [nvaccess/nvda] Python3 convert braille drivers (#9736)

 

I can test the hims, eurobraille, handy tech and baum drivers. Furthermore, I wrote most of the ALVA driver so will check that from a code point of view.



—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/9736?email_source=notifications&email_token=AB4AXECCQQ3NQ4UVKXR7NFLP2JNJRA5CNFSM4HXZP3J2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXT6P4Y#issuecomment-501737459> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AB4AXEBJPKFRYZHQPMU5N6LP2JNJRANCNFSM4HXZP3JQ> .

zstanecic commented:
Hi,

I can test this with my Esys 24, but I need thorough instructions how to use the branch py3 staging

 

From: Joseph Lee <notifications@github.com> 
Sent: Thursday, June 13, 2019 4:54 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Subject: Re: [nvaccess/nvda] Python3 convert braille drivers (#9736)

 

Hi, I can ask my BrailleNote to take part in testing this (since I also wrote code for Apex and QT support). Thanks.



From: Leonard de Ruijter <notifications@github.com <mailto:notifications@github.com> > 
Sent: Thursday, June 13, 2019 7:52 AM
To: nvaccess/nvda <nvda@noreply.github.com <mailto:nvda@noreply.github.com> >
Cc: Subscribed <subscribed@noreply.github.com <mailto:subscribed@noreply.github.com> >
Subject: Re: [nvaccess/nvda] Python3 convert braille drivers (#9736)



I can test the hims, eurobraille, handy tech and baum drivers. Furthermore, I wrote most of the ALVA driver so will check that from a code point of view.



—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/9736?email_source=notifications <https://github.com/nvaccess/nvda/pull/9736?email_source=notifications&email_token=AB4AXECCQQ3NQ4UVKXR7NFLP2JNJRA5CNFSM4HXZP3J2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXT6P4Y#issuecomment-501737459> &email_token=AB4AXECCQQ3NQ4UVKXR7NFLP2JNJRA5CNFSM4HXZP3J2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXT6P4Y#issuecomment-501737459> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AB4AXEBJPKFRYZHQPMU5N6LP2JNJRANCNFSM4HXZP3JQ> .




—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/9736?email_source=notifications&email_token=ACVCDEYF5WWVT2ZF2YVQKZ3P2JNQ5A5CNFSM4HXZP3J2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXT6WCA#issuecomment-501738248> , or mute the thread <https://github.com/notifications/unsubscribe-auth/ACVCDE5O5MAE5MZ5CG2DRZLP2JNQ5ANCNFSM4HXZP3JQ> .  <https://github.com/notifications/beacon/ACVCDE3EN6HMJTJQAKYTJPTP2JNQ5A5CNFSM4HXZP3J2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXT6WCA.gif>

zstanecic commented:
Hi,

As i said, i can test my Esys 24, 

How can I use these python 3 builds?



 

From: Leonard de Ruijter <notifications@github.com> 
Sent: Thursday, June 13, 2019 4:52 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Subject: Re: [nvaccess/nvda] Python3 convert braille drivers (#9736)

 

I can test the hims, eurobraille, handy tech and baum drivers. Furthermore, I wrote most of the ALVA driver so will check that from a code point of view.



—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/9736?email_source=notifications&email_token=ACVCDE5AZN2KKUM5ETF62SLP2JNJTA5CNFSM4HXZP3J2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXT6P4Y#issuecomment-501737459> , or mute the thread <https://github.com/notifications/unsubscribe-auth/ACVCDE4UIOFHG44WH2GVVMLP2JNJTANCNFSM4HXZP3JQ> .  <https://github.com/notifications/beacon/ACVCDE2R5QAH4ZZ35AIGDNLP2JNJTA5CNFSM4HXZP3J2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXT6P4Y.gif>

leonardder commented on source/braille.py:
```diff
@@ -1553,7 +1552,7 @@ class BrailleHandler(baseObject.AutoPropertyObject):
 	def __init__(self):
 		louisHelper.initialize()
-		self.display = None
+		self.display = None  # BrailleDisplayDriver
```
It looks like something like this is allowed:

```suggestion
		self.display: BrailleDisplayDriver = None 
```

leonardder commented on source/brailleDisplayDrivers/alva.py:
```diff
@@ -204,35 +205,48 @@ def terminate(self):
-		if data != b"\x1b": # Escape
+	def _ser6OnReceive(self, data: bytes):
+		"""
+		@type data: bytes
```
Does this make any sense if the type of data is already defined in the function signature?

feerrenrut replied:
Probably not, I will try to remove these if I notice them. That said, docstrings providing other useful information are still desired.

leonardder commented on source/brailleDisplayDrivers/alva.py:
```diff
@@ -204,35 +205,48 @@ def terminate(self):
+		if cmd == b"K":  # Input
+			self._handleInput(group=value[0], number=value[1])
+		elif cmd == b"E":  # Braille cell count
+			self.numCells = ord(value)  # this command only gets one byte
```
To be consistent with the line above, I think value[0] could be considered. May be we should come up with some standard here, i.e. always use ord for single bytes, or use index 0 for single bytes.

feerrenrut replied:
The pattern I have started to fall into with this is as you said, use ord when you know it is a single byte, and `__getitem__` when it is larger. This case meets that, the comment is not particularly insightful, at this point I was hoping to be able to confirm the sizes of all the commands (`cmd` values). But it is too time consuming in the scope of this PR.

leonardder commented on source/brailleDisplayDrivers/alva.py:
```diff
@@ -204,35 +205,48 @@ def terminate(self):
+		@type data: bytes
+		"""
 		reportID = data[0]
 		if reportID == ALVA_KEY_REPORT:
```
This is going to fail I think. data[0] resulted in an int and you're comparing against a single byte in a bytes object.

feerrenrut replied:
Yes, good catch. int to bytes comparisons are hard to catch, the linter seems to miss them, silly really.

leonardder commented on source/brailleDisplayDrivers/alva.py:
```diff
@@ -258,44 +272,59 @@ def _handleInput(self, group, number):
+			cellsToWrite = cellBytes[offset:offset+ALVA_BRAILLE_OUTPUT_MAX_SIZE]
+			data = b"".join([
+				ALVA_BRAILLE_OUTPUT_REPORT,
+				intToByte(offset),
```
I believe there is a [proposal](https://www.python.org/dev/peps/pep-0467/) planned for python 3.9 for this to be called bchr. We could anticipate for this.

feerrenrut replied:
I don't think we should name this `bchr` if that is what you are suggesting, that could lead to an unexpected change in behaviour when we upgrade.

leonardder commented on source/brailleDisplayDrivers/alva.py:
```diff
@@ -258,44 +272,59 @@ def _handleInput(self, group, number):
+		])
+		self._ser6SendMessage(b"B", value)
+
+	def display(self, cells: List[int]):
```
I think now we can, we should consider making this a bytes or bytearray object. In most if not all cases, one of the first things that happens in this method is converting a list to a byte sequence.

feerrenrut replied:
I understand, however, I think it's best to start with the smallest change set. This isn't really a change from how it worked with python 2, and it because it is quite self contained it should be fairly easy to do after this PR.

leonardder commented on source/brailleDisplayDrivers/alva.py:
```diff
@@ -258,44 +272,59 @@ def _handleInput(self, group, number):
+		"""
 		# cells will already be padded up to numCells.
-		cells = b"".join(map(chr, cells))
+		cellBytes = b"".join(
```
See? It is just ridiculous to do this in every single driver.

leonardder commented on source/brailleDisplayDrivers/alva.py:
```diff
@@ -304,14 +333,17 @@ def _handleTime(self, timeStr):
-	def _syncTime(self, dt):
+	def _syncTime(self, dt: datetime.datetime):
 		log.debug("Synchronizing braille display date and time...")
 		timeList = [
```
Now, this really looks like a case where we might want to use a bytearray instead.

LeonarddeR reviewed (commented):
Here are my comments for the diff until and up to the alva driver.

feerrenrut commented:
@zstanecic and any other interested in testing, thank you for your interest, we will make an announcement when we would like to invite testing and will provide a build. For now this is very much a work in progress.

leonardder commented:
cc @bramdbb

leonardder commented:
cc @bramd

leonardder commented:
I've been thinking about the several cases where you know do something like b"".join([arg, int2byte(arg2)]).

I think this can be really error prone if you don't have the proper linting tooling.

Have you considered something like this?

```
def createPacket(*args):
	packet = bytearray()
	for arg in args:
		if isinstance(arg, int):
			packet.append(arg)
		elif isinstance(arg, bool):
			packet.extend(boolToByte(arg))
		elif isinstance(arg, collections.abc.ByteString):
			packet.extend(arg)
		else:
			raise TypeError
	return packet
```

leonardder commented on source/brailleDisplayDrivers/baum.py:
```diff
@@ -18,21 +21,21 @@
-BAUM_JOYSTICK_KEYS = "\x34"
-BAUM_DEVICE_ID = "\x84"
-BAUM_SERIAL_NUMBER = "\x8A"
+ESCAPE = b"\x1b"
```
I think that many drivers would benefit from a new module (e.g. hwIo.constants) containing stuff like ESCAPE, STX, ETX, etc.

feerrenrut commented:
The usage of that will end up like:
```
createPacket(
  data[2], # an int
  data[3:6], # bytes
  b"/x67", # bytes
  "/75", # oops a unicode string, linting wont catch me.
)
```

I think that it hides the problem. At least with `b"".join` you know that everything in the list hast to be bytes (or 'bytes-like'). With a function like this, you could easily be convinced into thinking it automatically converted a case that it didn't. That said, I think the `b"".join` syntax is quite ugly.

leonardder commented on source/brailleDisplayDrivers/baum.py:
```diff
@@ -131,19 +135,41 @@ def terminate(self):
-				arg=arg.replace(ESCAPE, ESCAPE * 2)))
+			arg = arg.replace(ESCAPE, ESCAPE * 2)
+			data = b"".join([
+				b"\x1b",
```
This is `ESCAPE`

leonardder commented on source/brailleDisplayDrivers/baum.py:
```diff
@@ -131,19 +135,41 @@ def terminate(self):
+			])
+			self._dev.write(data)
 
 	def _onReceive(self, data):
```
```suggestion
	def _onReceive(self, data: bytes -> None):
```

leonardder commented on source/brailleDisplayDrivers/baum.py:
```diff
@@ -162,11 +188,17 @@ def _onReceive(self, data):
 		self._handleResponse(command, arg)
 
 	def _handleResponse(self, command, arg):
```
```suggestion
	def _handleResponse(self, command: bytes, arg: bytes -> None):
```

leonardder commented on source/brailleDisplayDrivers/baum.py:
```diff
@@ -162,11 +188,17 @@ def _onReceive(self, data):
+		:type arg: bytes
+		"""
 		if command == BAUM_CELL_COUNT:
+			# Assumption: BAUM_CELL_COUNT command has a single byte unsigned argument.
```
A cell count usually is not much more than 80.

leonardder replied:
Within the scope of braille displays, a cell count is pretty self explanatory. I think you can leave the comment as is.

leonardder commented on source/brailleDisplayDrivers/baum.py:
```diff
@@ -162,11 +188,17 @@ def _onReceive(self, data):
 		elif command == BAUM_DEVICE_ID:
 			# Short ids can be padded with either nulls or spaces.
-			self._deviceID = arg.rstrip("\0 ")
+			self._deviceID = arg.rstrip(b"\0 ")
```
I think we should save this one as a string.

```suggestion
			self._deviceID = arg.rstrip(b"\0 ").decode("latin-1") # Assumption
```

leonardder commented on source/brailleDisplayDrivers/brailliantB.py:
```diff
@@ -21,18 +24,18 @@
-MSG_DISPLAY = "\x02"
-MSG_KEY_DOWN = "\x05"
-MSG_KEY_UP = "\x06"
+HEADER = b"\x1b"
```
Interesting to see how many protocols start with ESCAPE.

leonardder commented on source/brailleDisplayDrivers/brailliantB.py:
```diff
@@ -182,11 +195,11 @@ def _serHandleResponse(self, msgId, payload):
-	def _hidOnReceive(self, data):
+	def _hidOnReceive(self, data: bytes):
 		rId = data[0]
 		if rId == HR_KEYS:
```
Woops, another fail of the linter I presume?

feerrenrut replied:
Yeah, it doesn't pick up these cases at all, its legal to allow comparison of int and bytes, but it will always be false. Given that you'd think there would be a warning though! I don't think I realised that the linter missed it until one of the later drivers, so I probably wasn't as careful here,  though it is easy to miss. Drivers are done one per commit, so hopefully I caught these in the later ones. I'm still only about half way through!

leonardder commented on source/brailleDisplayDrivers/ecoBraille.py:
```diff
@@ -67,41 +69,53 @@ class ecoTypes:
+	if len(msg) < 9:
 		return ecoTypes.TECO_80 # Needed to restart NVDA with Ecoplus
-	msg = struct.unpack('BBBBBBBBB', msg)
+	msgUnpacked: Tuple[int, ...] = struct.unpack(b'BBBBBBBBB', msg)
```
Actually, I think you can safely remove this unpack.

leonardder commented on source/brailleDisplayDrivers/ecoBraille.py:
```diff
@@ -67,41 +69,53 @@ class ecoTypes:
+def eco_in(dev: serial.Serial) -> int:
 	try:
-		msg = struct.unpack('BBBBBBBBB', msg)
+		msg: Tuple[int, ...] = struct.unpack(b'BBBBBBBBB', dev.read(9))
```
Also here, indexing the bytes object just returns ints, so we no longer have to unpack.

LeonarddeR reviewed (commented):
Comments up to and including ecoBraille

leonardder commented:
unfortunately, there is no in line way to populate a bytearry with groups of bytes, only single bytes.

leonardder commented on source/brailleDisplayDrivers/eurobraille.py:
```diff
@@ -188,7 +196,9 @@ def __init__(self, port="Auto"):
 				self._sendPacket(EB_SYSTEM, EB_SYSTEM_IDENTITY)
 				# Make sure visualisation packets are disabled, as we ignore them anyway.
-				self._sendPacket(EB_VISU, EB_VISU_DOT, '0')
+				# py3 review: this is currently a zero char (0b110001 not 0b0).
```
The eurobraille protocol uses real number characters as boolean values, so 0 (0x30) and 1 (0x31)

leonardder commented on source/brailleDisplayDrivers/eurobraille.py:
```diff
@@ -284,51 +301,51 @@ def _handleAck(self, frame):
+	def _handleKeyPacket(self, group: bytes, data: bytes):
 		if group == EB_KEY_USB_HID_MODE:
-			self._hidKeyboardInput = bool(int(data))
+			self._hidKeyboardInput = boolToBytes(data)
```
I don't think this does the same. bool(int(data)) will be False if data == b"0", where byte2bool will use the ordinal of the byte to check the boolean value. See my comment above regarding the eurobraille protocol using real numbers for boolean values.

leonardder commented on source/brailleDisplayDrivers/eurobraille.py:
```diff
@@ -284,51 +301,51 @@ def _handleAck(self, frame):
 			return
-		if group == EB_KEY_INTERACTIVE and data[0]==EB_KEY_INTERACTIVE_REPETITION:
-			log.debug("Ignoring routing key %d repetition"%(ord(data[1])-1))
+		if group == EB_KEY_INTERACTIVE and data[0] == EB_KEY_INTERACTIVE_REPETITION:
```
Please have look at every case where data[0] (which translates to an int) is used.

leonardder commented on source/brailleDisplayDrivers/eurobraille.py:
```diff
@@ -337,18 +354,18 @@ def _handleKeyPacket(self, group, data):
 			STX,
-			chr((packetSize>>8)&0xff),
-			chr(packetSize&0xff),
+			chr((packetSize >> 8) & 0xff),
```
chr should be int2byte

leonardder commented on source/brailleDisplayDrivers/eurobraille.py:
```diff
@@ -357,34 +374,51 @@ def _sendPacket(self, packetType, packetSubType, packetData=b""):
+			packetType=EB_BRAILLE_DISPLAY,
+			packetSubType=EB_BRAILLE_DISPLAY_STATIC,
+			packetData=b"".join([
+				intToBytes(cell) for cell in cells
```
Thinking more about this, shouldnt bytes(cells) be sufficient here?

leonardder commented on source/brailleDisplayDrivers/eurobraille.py:
```diff
@@ -357,34 +374,51 @@ def _sendPacket(self, packetType, packetSubType, packetData=b""):
+			# py3 review: line was: =str(int(state))
+			# This would result in b'1' == 49 == 0b110001
+			# Instead the output will be b'\x01 == 0b1 (eg: unsigned int(1))
+			packetData=boolToBytes(state)
```
Please revert this as per above comments.

LeonarddeR reviewed (changes requested):
Comments for the eurobraille driver

feerrenrut committed:
Python3: Convert hedoMobilLine driver

feerrenrut committed:
Convert hedoProfiLine driver

leonardder commented on source/brailleDisplayDrivers/handyTech.py:
```diff
@@ -773,21 +775,21 @@ def _handleInputStream(self, htPacketType, stream):
 		else:
 			serPacketOrd = ord(htPacketType)
-			if isinstance(self._model, OldProtocolMixin) and serPacketOrd&~KEY_RELEASE_MASK < HT_PKT_EXTENDED:
+			if isinstance(self._model, OldProtocolMixin) and serPacketOrd&~KEY_RELEASE_MASK < ord(HT_PKT_EXTENDED):
```
Good catch!

leonardder commented on source/hwIo.py:
```diff
@@ -89,18 +91,24 @@ def waitForRead(self, timeout):
+		"""
+		@type data: bytes
+		"""
+		if not isinstance(data, bytes):
```
I think we should use collections.abc.ByteString here, so that bytearray is also allowed as a valid type. It is fully compatible.

feerrenrut replied:
Perhaps, but I think it's outside of the scope of this PR to go down that road. EG finding the optimal types to allow the greatest flexibility. This PR should just be concerned with getting it working. If someone shows that it causes some perf concern to do `bytes(myByteArray)`, or taking a `bytes` type only is too inflexible for some given case we can look at changing it.

leonardder replied:
Note that before this pr, the write function just accepted everything you threwat it. You're now introducing this explicit check for bytes. I won't insist, but I really believe that it is too strict, especially given bytearray is fully compatible with bytes and they are both handled by the ByteString abstract class.

feerrenrut replied:
I reduced the types it could take to simplify verifying correctness. Taking multiple types introduces uncertainty, for instance taking an `int`, the code must guess how to convert it (number of bytes, byte order, signed), these are device specific and should be decided in the driver. There is certainly an argument for accepting a `bytearray`, however I don't see that it's necessary, and doesn't greatly impact the code. While it's tempting to try to build a general purpose library, it's much easier to get the known cases correct if we focus on those only.

> especially given bytearray is fully compatible with bytes 

It should be noted that a `bytearray` is not convertible by `ctypes.create_string_buffer` and will not be automatically converted to a bytes object.

leonardder commented on source/hwIo.py:
```diff
@@ -89,18 +91,24 @@ def waitForRead(self, timeout):
-		buf.raw = data
+		# usage of self._writeSize could result in a size mismatch with data?
+		if size != len(data):
+			log.debugWarning(u"Mismatch in buffer size writing to braille device.")
```
I think we only want this if _isDebug()

feerrenrut replied:
Are you suggesting changing the log level to debug, `len(data)` has a high perf cost, or that this logging is too noisy and we should suppress it unless in a special braille debug mode?

leonardder replied:
hwIo has a _isDebug function that returns True of hwIo specific debugging is True, False otherwise. I think this information should stay under the hwIo debug flag.

Having said that actually, most HID drivers just throw a packet at write that isn't the writeSize, so I'm afraid this will lead to many, many debug warnings. I think it is perfectly valid for drivers to rely on NVDA to send a packet to the device that is of the right size.

It would be good to know what @jcsteh thinks about this, since He's the initial author of hwIo. But my advice would be that if we really want to log this, it should be at debug and behind the hwIo flag. Probably use logHandler.log.isEnabledFor, indeed.

leonardder commented on source/hwIo.py:
```diff
@@ -304,34 +323,42 @@ def getFeature(self, reportId):
-		length = len(report)
-		buf = ctypes.create_string_buffer(length)
-		buf.raw = report
+		buf = ctypes.create_string_buffer(report)
```
I think this buffer is no 1 byte longer than expected, as a string buffer will probably contain a terminating character.

feerrenrut replied:
Good catch, I can pass the size along with it to stop this.

leonardder commented on source/hwIo.py:
```diff
@@ -304,34 +323,42 @@ def getFeature(self, reportId):
-		length=len(report)
-		buf=ctypes.create_string_buffer(length)
-		buf.raw=report
+		buf = ctypes.create_string_buffer(report)
```
Same as above, I think this results in a longer buffer than expected.

LeonarddeR reviewed (changes requested):
Looks like I'm finished for now.

feerrenrut committed:
Convert hims driver

Some extra annotations in braille.py and bdDetect.py.
Noticed that these files (braille.py and bdDetect.py) require lots more work.

feerrenrut committed:
Convert lilli driver

leonardder commented on source/brailleDisplayDrivers/lilli.py:
```diff
@@ -66,31 +69,36 @@ def terminate(self):
-				key=lilliDll.ReadBuf()
+				# Python 3: review required
+				# The code seems to assume this returns an int.
+				# I haven't confirmed this.
```
I think this assumption is safe. By default, every function from a windll type dll returns a long type.

LeonarddeR reviewed (commented):
I'm afraid we have no reached the point where this pr is growing to big. It took me around 10 minutes to send my review comments, just because the vbuf is so huge.
Actually, both hedo drivers are ok, so that's why I didn't write comments there.

bramd commented:
@feerrenrut Which linter did you use for type checking? There seem to be a few of them now.

bramd commented on source/brailleDisplayDrivers/freedomScientific.py:
```diff
@@ -287,28 +296,32 @@ def _onReceive(self, data):
 		if self.isUsb:
 			data = BytesIO(data)
-			packetType = data.read(1)
+			packetType: bytes = data.read(1)
```
Since this is just one byte, do we have a better type declaration for that?

feerrenrut replied:
No, I don't think so. You could do `data.read(1)[0]` or `ord(data.read(1))` to take it as an int. Considering we then compare to several bytes values, it's best not to change the type. You can't compare bytes to int values directly.

bramd commented on source/brailleDisplayDrivers/freedomScientific.py:
```diff
@@ -342,15 +355,26 @@ def _handlePacket(self, packetType, arg1, arg2, arg3, payload):
+			manuBytes = payload[INFO_MANU_START:INFO_MANU_END].replace(
+					FS_BYTE_NULL, b""
+			)
+			self._manufacturer = manuBytes.decode()
```
I think latin-1 would match Python 2's behavior. However, assuming UTF-8 might not be a bad idea if a display will have a name including UTF-8 characters in the future. As far as I know, the first characters of latin-1/utf-8 are the same?

feerrenrut replied:
If the first characters are the same, then can we just leave it as UTF-8? Would it be more future proof?

bramd reviewed (commented):
I can't get this branch to build yet, so didn't do any real testing.

feerrenrut committed:
Convert papenmeier driver

feerrenrut committed:
Move Input gesture functions in papenmeier driver

feerrenrut committed:
Convert papenmeier serial driver

feerrenrut commented on source/brailleDisplayDrivers/freedomScientific.py:
```diff
@@ -256,29 +261,33 @@ def _sendPacket(self, packetType, arg1=FS_BYTE_NULL, arg2=FS_BYTE_NULL, arg3=FS_
-			if type(arg) == int:
-				return int2byte(arg)
-			return arg
+		def handleArg(arg: bytes) -> bytes:
```
`_sendPacket` only has one argument that was being passed as an int.  It adds uncertainty to the size and type of the arguments at several layers. To convert int's at the last moment means making a general assumption about the byte representation of the int. Are all future cases covered by representing an int with a single byte, unsigned? I prefer it's clear that this conversion must take place.

feerrenrut commented on source/brailleDisplayDrivers/handyTech.py:
```diff
@@ -247,16 +251,16 @@ def handleTime(self, timeStr):
+	def syncTime(self, dt: datetime.datetime):
 		log.debug("Synchronizing braille display date and time...")
 		# Setting the time uses a swapped byte order for the year.
 		timeList = [
```
How would it be different from what we have at the moment?

leonardder replied:
We could throw a bytearray at sendExtendedPacket directly, we do no longer have to convert it to bytes explicitly.

feerrenrut commented:
> @feerrenrut Which linter did you use for type checking? There seem to be a few of them now.

@bramd I am using whatever is built into pycharm. We are looking at integrating flake8 with the NVDA build process once I am done with these drivers. But more on that later :)

feerrenrut committed:
Review actions

feerrenrut commented:
I think I have handled all the review actions. I will take one more pass over this looking for errors that I only became aware of later in the process. But if everyone is happy that these changes result in equal behaviour, then I think we should merge this.

leonardder commented:
Before merging, I'd like to have a quick test with at least the handy 
tech, eurobraille, hims and baum drivers. I can do that next monday at 
the @babbagecom office if you'd like, so that shouldn't cause any delay 
while hopefully improving the code quality and stability.

josephsl commented:
Hi, unit tests are passing (if I tell Unittest to skip the braille gestures identifier test, it works). I’ll do some more runtime tests with my Apex (and if it consents to it, a new BrailleNote Touch Plus) this weekend. Thanks.

 

From: Leonard de Ruijter <notifications@github.com> 
Sent: Friday, June 21, 2019 10:54 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] Python3 convert braille drivers (#9736)

 

Before merging, I'd like to have a quick test with at least the handy 
tech, eurobraille, hims and baum drivers. I can do that next monday at 
the @babbagecom office if you'd like, so that shouldn't cause any delay 
while hopefully improving the code quality and stability.



—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/9736?email_source=notifications&email_token=AB4AXEF7EAVWMAKYXHLQGTTP3UIUVA5CNFSM4HXZP3J2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYJEWOQ#issuecomment-504515386> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AB4AXEHE2JLI5TJZHEIO5E3P3UIUVANCNFSM4HXZP3JQ> .

jcsteh commented:
The whole point of writeSize was precisely that HID reports must be a
specific size, but not all packets use the full size. So, I think this
warning should not be logged at all.

josephsl commented:
Hi,

Braille input fails because braille input module still calls unichr. This is after testing this branch with my BrailleNote Apex. I bet this will be addressed by another PR - if not, I think it would be great if this can be addressed.

Thanks.

feerrenrut commented:
> Braille input fails because braille input module still calls unichr. 

Thanks @josephsl I'll look at this.

feerrenrut commented:
> The whole point of writeSize was precisely that HID reports must be a specific size, but not all packets use the full size. So, I think this warning should not be logged at all.

@jcsteh I've been looking at [the docs](https://docs.microsoft.com/en-gb/windows/desktop/api/fileapi/nf-fileapi-writefile) and haven't noticed this specified in relation to WriteFile

There is this comment in the code:
```
# Reading any less than caps.InputReportByteLength is an error.
# On Windows 7, writing any less than caps.OutputReportByteLength is also an error.
```
But it doesn't tell us what error or where to get information on it. I'll look into this more on Monday, perhaps there is more information in the original commit messages / PR.

My concerns:
- A `writeSize` of less than the buffer is going to result some data not being sent to the device. I suspect this is not the case, it would likely stop devices from working.
- A `writeSize` of more than the buffer is going to result in reading past the end of the buffer, sending garbage. 

If write size must be a fixed size, then we should be altering the size of the buffer to match and padding with something, or doing multiple writes.

leonardder commented on source/brailleDisplayDrivers/papenmeier_serial.py:
```diff
@@ -28,31 +27,33 @@
+	ret = bytearray([
+		STX,
+		ord('S'),
+		offset // 256,
```
Wow, this really looks like poorman's bit shifting

```suggestion
		offset >> 8,
```

We can also leave it as is, though, but I consider it pretty missleading.
Note that I also ensured the validity of this change:

```
for i in range(0x1000000):
	assert i // 256 == i >> 8
```

leonardder commented on source/brailleDisplayDrivers/papenmeier_serial.py:
```diff
@@ -28,31 +27,33 @@
+		STX,
+		ord('S'),
+		offset // 256,
+		offset % 256,
```
```suggestion
		offset & 0xff,
```

Test case:

```
for i in range(0x1000000):
	assert i % 256 == i & 0xff
```

leonardder commented on source/brailleDisplayDrivers/papenmeier_serial.py:
```diff
@@ -28,31 +27,33 @@
+		offset // 256,
+		offset % 256,
+		0,
+		d2 % 256,
```
```suggestion
		d2 & 0xff
```

leonardder commented on source/brailleDisplayDrivers/papenmeier_serial.py:
```diff
@@ -9,15 +9,14 @@
 
```
Could you find out whether this driver is still in use by some people? I think that it is very unlikely that this will ever be tested before a release, due to the age of these devices.

leonardder commented on source/brailleDisplayDrivers/papenmeier_serial.py:
```diff
@@ -182,11 +182,11 @@ def _handleKeyPresses(self): #called by the keycheck timer
-			if(len(data) == 10 and data[1]=='K'):
-				pos = ord(data[2])*256+ord(data[3])
+			if len(data) == 10 and data[1] == ord('K'):
+				pos = data[2] * 256 + data[3]
```
```suggestion
				pos = data[2] << 8 | data[3]
```

feerrenrut replied:
Probably better to use `int.from_bytes` to convert this, the intention is clearer. That said, I'm going to leave it as is. It's not required to change to Python 3.

leonardder commented on source/brailleDisplayDrivers/papenmeier_serial.py:
```diff
@@ -182,11 +182,11 @@ def _handleKeyPresses(self): #called by the keycheck timer
-				pos = ord(data[2])*256+ord(data[3])
+			if len(data) == 10 and data[1] == ord('K'):
+				pos = data[2] * 256 + data[3]
 				pos = (pos-768)/3
```
It looks like this can ever be negative. I'm not sure whether this is expected, but I don't think so. May be we should add a comment to this code that there are some code quality concerns here.

leonardder commented on source/brailleDisplayDrivers/papenmeier_serial.py:
```diff
@@ -253,7 +253,13 @@ class InputGesture(braille.BrailleDisplayGesture):
-	def __init__(self, keyindex, pressed, keys, driver):
+	def __init__(
+			self,
+			keyindex: Union[int, None],
```
```suggestion
			keyindex: Optional[int],
```

Also the cases below

jcsteh commented:
This behaviour of WriteFile is specific to HID devices. Moreover, it isn't well documented and the behaviour differs between versions of Windows. I don't recall the exact error code returned from WriteFile. I do recall having to figure it out through testing with a device; i.e. the docs weren't enough.

It's reasonable to say that writing more than writeSize is an error. I think throwing an exception in that case is reasonable. Certainly, no drivers (current or future) should be doing this.

As for a buffer less than writeSize, we're not reading past the end of the buffer because we're creating a buffer of writeSize. When we put the Python string into the buffer, ctypes only reads as far as the length of the string, since it knows the length. I'm fairly sure ctypes zeroes out buffers it creates (resulting in zero padding). Even if it doesn't, the device should ignore the remaining bytes; the protocol specifies the length of each report and the remaining bytes are just to satisfy the silly Windows requirement of having to write to the maximum report size.

Commit:
Fixes/improvements to hwIo and fixes to handy tech driver

bdorer commented:
Leonard comented among other coments:
In source/brailleDisplayDrivers/papenmeier_serial.py:
> @@ -9,15 +9,14 @@
Could you find out whether this driver is still in use by some people? I 
think that it is very unlikely that this will ever be tested before a 
release, due to the age of these devices.
I know serveral German persons who use such displays. May 
@michaelDCurran could provide some statistic.

michaelDCurran commented:
We see a maximum of 8 users using papenmeier_serial over the last 30 days.

Of course this only counts users who opted in to stats gathering, so the 
number could be 30% higher, perhaps 10 or 11 users. Plus, it only tracks 
users from 2018.3 onwards.

leonardder commented:
I'm merging in current threshold_py3_staging, so we can check appveyor status for this

Commit:
Merge remote-tracking branch 'origin/threshold_py3_staging' into python3-convertBrailleDrivers

feerrenrut commented on source/hwIo.py:
```diff
@@ -32,15 +34,22 @@ class IoBase(object):
-	def __init__(self, fileHandle, onReceive, writeFileHandle=None, onReceiveSize=1, writeSize=None):
+	def __init__(
+			self,
+			fileHandle: Union[ctypes.wintypes.HANDLE, Any],
```
It means that we are not getting proper type checking, but in the mean time it is to document that there is another type that is accepted. I actually don't know what type it is, it is the `_port_handle` in the serial implementation.

leonardder replied:
I"m actually pretty sure that it will be just an int, but will be able to verify that tomorrow.

feerrenrut replied:
I think I will just remove `Any`, I assume `_port_handle` it has to be can substitute for a `ctypes.wintypes.HANDLE` anyway.

leonardder replied:
I verified that _port_handle is an int. I dont think that's really interchangeable with ctypes.wintypes.HANDLE, which is actually just a c_void_p, so anything non ctypes that takes an int can't handle it. ctypes can interpret them both though.

feerrenrut committed:
Create a buffer manually for writes to HID devices

By creating the buffer manually, we can ensure that data smaller than OutputReportByteLength
will be padded with null bytes. I tested ctypes marshalling by creating
a C++ dll, loading it and calling it with ctypes.

Small refactor, only hwIo.HID devices were using _writeSize. hwIo.Bulk
took the parameter, but it was only set to 0 by the hims driver. This
failed the Truth test, and len(data) was used in this case.

The refactor allows the HID complexity to stay with the HID class.
If it becomes necessary to support buffers larger than OutputReportByteLength
further refactoring will become necessary.

feerrenrut committed:
Merge branch 'driverFixes' of https://github.com/BabbageCom/nvda into python3-convertBrailleDrivers

feerrenrut commented on source/brailleDisplayDrivers/brailleNote.py:
```diff
@@ -166,25 +167,31 @@ def _describe(self):
+			self,
+			command: int,
+			arg: int,
+			arg2: Optional[str] = None
```
No, this is used to construct an input gesture. It comes from `_qtKeys` which is a mapping (dictionary) of  `int` (I assume button codes) to names of keys.

feerrenrut committed:
Further review actions:

Use Optional instead of Union[sometype, None]

Checked all usages of:
- ord()
- bytes()
- bytearray()
- str()

Found and fixed several issues.

Commit:
Baum: Change if to elif, add more debug info

feerrenrut commented:
I've just re-requested reviews from @leonardder and @bramd.

Just a note, we would like to merge this as soon as we can to start getting broader testing. Please focus reviewing efforts on looking for bugs / mistakes.

leonardder commented:
I have tested the following driver/display combinations over USB:

Driver | Display
----|----
Handy Tech | Modular Evolution 88 and Active Braille
alva | Alva BC640
Papenmeier | I recall an EL40C, not sture about the type number though
Eurobraille | Esys 40 with firmware 3.16
Hims | Braille Edge
Baum | Vario Ultra 40

LeonarddeR reviewed (approved):
I looked at the diff from last review actions, and they look ok.

michaelDCurran commented:
Was the idea for this pr to also convert brailleInput.py? @feerrenrut 
requested me not to touch brailleInput.py when fixing unichr issues but 
in this pr I'm getting the following traceback when typing on an Orbit 
Reader braille display:

```

Traceback (most recent call last):
   File "scriptHandler.py", line 189, in executeScript
     script(gesture)
   File "globalCommands.py", line 2059, in script_braille_dots
     brailleInput.handler.input(gesture.dots)
   File "brailleInput.py", line 247, in input
     if self._translate(endWord):
   File "brailleInput.py", line 129, in _translate
     data = u"".join([unichr(cell | LOUIS_DOTS_IO_START) for cell in 
self.bufferBraille[:pos]])
   File "brailleInput.py", line 129, in <listcomp>
     data = u"".join([unichr(cell | LOUIS_DOTS_IO_START) for cell in 
self.bufferBraille[:pos]])
NameError: name 'unichr' is not defined
```


On 6/27/2019 11:52 PM, Leonard de Ruijter wrote:
>
> *@leonardder* approved this pull request.
>
> I looked at the diff from last review actions, and they look ok.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/nvaccess/nvda/pull/9736?email_source=notifications&email_token=ADL7THLJ5PDFUWJFSTMQQ3LP4TA3NA5CNFSM4HXZP3J2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB43K3QY#pullrequestreview-255241667>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/ADL7THIVW6532G3MUJI6543P4TA3NANCNFSM4HXZP3JQ>.
>

michaelDCurran commented:
Successfully used an Orbit Reader to read and navigate with NVDA, in both bluetooth and USB HID. Was also able to type braille input with #9835 merged into this branch.

michaelDCurran commented:
Successfully read / navigated / typed on Baum Refreshabraille 18 with both USB and blutooth.

feerrenrut commented:
> Was the idea for this pr to also convert brailleInput.py?

Yes, actually, it seems I missed it. Perhaps we merge these two PR's then I'll take another look through `brailleInput.py` looking for any other issues.

bramd commented:
I tested the following displays over USB:

* Freedom Scientific Focus blue
* HumanWare Brailliant BI 40 (serial and HID)

These work fine after applying the fixes in #9839

feerrenrut committed:
add type hints to brailleInput.py

feerrenrut committed:
Merge remote-tracking branch 'origin/threshold_py3_staging' into python3-convertBrailleDrivers
